File: Sample Apps/Objective C/SampleApp/MonitoringViewController.m
//
//  MonitoringViewController.m
//  SampleApp
//
//  Created by Nir Lachman on 12/02/2018.
//  Copyright © 2018 LivePerson. All rights reserved.
//

#import "MonitoringViewController.h"
#import <LPMonitoring/LPMonitoring.h>
#import <LPInfra/LPInfra.h>
#import <LPMessagingSDK/LPMessagingSDK.h>

NSString * const consumerID = @"CONSUMER_ID"; // REPLACE THIS!
NSString * const appInstallID = @"APP_INSTALL_ID"; // REPLACE THIS!
@interface MonitoringViewController ()
@property (nonatomic, strong)  NSString * _Nullable  pageId;
@property (nonatomic, strong)  LPCampaignInfo * _Nullable  campaignInfo;
@end

@implementation MonitoringViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

// MARK: IBAction
/// Init Messaging SDK with brandID (account number) and LPMonitoringInitParams (For monitoring)
- (IBAction)initSdkClicked:(id)sender {
    [[self view] endEditing:YES];
    NSString *account = self.accountTextField.text;
    if (account.length > 0) {
        LPMonitoringInitParams *monitoringInitParams = [[LPMonitoringInitParams alloc] initWithAppInstallID: account];
        NSError *error = nil;
        [[LPMessagingSDK instance] initialize:account monitoringInitParams:monitoringInitParams error:&error];
        if (error) {
            NSLog(@"LPMessagingSDK Initialize Error: %@",error);
            return;
        }
    }
}

/// Get Engagement clicked selector
/// Send new Get Engagement using LPMonitoringAPI
/// CampaignInfo will be saved in the response in order to be able to use routed conversation in Messaging, using the MessagingSDK
- (IBAction)getEngagementClicked:(id)sender {
    [[self view] endEditing:YES];
    
    NSArray *entryPoints = @[@"tel://972737004000",
                             @"http://www.liveperson.com",
                             @"sec://Sport",
                             @"lang://Eng"];
    NSArray *engagementAttributes = @[
                                      @{@"type": @"purchase", @"total": @20.0},
                                      @{@"type": @"lead", @"lead": @{@"topic": @"luxury car test drive 2015", @"value": @22.22, @"leadId": @"xyz123"}},
                                      ];
    
    LPMonitoringParams *monitoringParams = [[LPMonitoringParams alloc] initWithEntryPoints:entryPoints engagementAttributes:engagementAttributes pageId:NULL];
    __weak MonitoringViewController *weakSelf = self;
      LPMonitoringIdentity *identity = [[LPMonitoringIdentity alloc] initWithConsumerID:consumerID
                                                                                 issuer:@""];
      [[LPMonitoringAPI instance] getEngagementWithIdentities:@[identity] monitoringParams:monitoringParams completion:^(LPGetEngagementResponse * _Nonnull getEngagementResponse) {
          weakSelf.pageId = getEngagementResponse.pageId;
          if (getEngagementResponse.engagementDetails.count > 0) {
              weakSelf.pageId = getEngagementResponse.pageId;
              LPEngagementDetails *engagementDetails = getEngagementResponse.engagementDetails[0];
              NSInteger campaignId = engagementDetails.campaignId;
              NSInteger engagementId = engagementDetails.engagementId;
              NSString *contextId = engagementDetails.contextId;
              NSString *sessionId = getEngagementResponse.sessionId;
              NSString *visitorId = getEngagementResponse.visitorId;
              if (campaignId > 0 && engagementId > 0 && contextId != nil) {
                  weakSelf.campaignInfo = [[LPCampaignInfo alloc] initWithCampaignId:campaignId
                                                                        engagementId:engagementId
                                                                           contextId:contextId
                                                                           sessionId:sessionId
                                                                           visitorId:visitorId];
              }
          }
      } failure:^(NSError * _Nonnull error) {
          weakSelf.campaignInfo = nil;
          NSLog(@"get engagement error: %@", error.localizedDescription);
      }];
}

/// Send SDE clicked selector
/// Send new SDE using LPMonitoringAPI
/// PageID in the response will be saved for future request for SDE
- (IBAction)sendSdeClicked:(id)sender {
    [[self view] endEditing:YES];
    
    NSArray *entryPoints = @[@"tel://972737004000",
                             @"http://www.liveperson.com",
                             @"sec://Sport",
                             @"lang://Eng"];
    NSArray *engagementAttributes = @[
                                      @{@"type": @"purchase", @"total": @20.0},
                                      @{@"type": @"lead", @"lead": @{@"topic": @"luxury car test drive 2015", @"value": @22.22, @"leadId": @"xyz123"}},
                                      ];
    
    __weak MonitoringViewController *weakSelf = self;
    LPMonitoringParams *monitoringParams = [[LPMonitoringParams alloc] initWithEntryPoints:entryPoints
                                                                      engagementAttributes:engagementAttributes
                                                                                    pageId:@"pageId"];
    LPMonitoringIdentity *identity = [[LPMonitoringIdentity alloc] initWithConsumerID:consumerID
                                                                               issuer:@"BrandIssuer"];
    [[LPMonitoringAPI instance] sendSDEWithIdentities:@[identity] monitoringParams:monitoringParams completion:^(LPSendSDEResponse * _Nonnull sendSdeResponse) {
        weakSelf.pageId = sendSdeResponse.pageId;
    } failure:^(NSError * _Nonnull error) {
        weakSelf.campaignInfo = nil;
        NSLog(@"send sde error: %@", error.localizedDescription);
    }];
}

/// Show Conversation clicked selector
/// Show conversation in MessagingSDK will use the saved CampaignInfo, if received, from the Get Engagement Request
- (IBAction)showConversationWithCampaignClicked:(id)sender {
    [[self view] endEditing:YES];
    
    if (self.campaignInfo != nil && self.accountTextField.text.length > 0) {
        id <ConversationParamProtocol> conversationQuery = [[LPMessagingSDK instance] getConversationBrandQuery:self.accountTextField.text campaignInfo:self.campaignInfo];
        LPConversationViewParams *conversationViewParams = [[LPConversationViewParams alloc] initWithConversationQuery:conversationQuery
                                                                                               containerViewController:nil
                                                                                                            isViewOnly:NO
                                                                                       conversationHistoryControlParam:nil];
        [[LPMessagingSDK instance] showConversation:conversationViewParams authenticationParams:nil];
    }
}

/// Logout clicked selector
/// Logout Messaging SDK - all the data will be cleared
- (IBAction)logoutClicked:(id)sender {
    [[self view] endEditing:YES];
    [[LPMessagingSDK instance] logoutWithCompletion:^{
        
    } failure:^(NSError * _Nonnull error) {
        
    }];
}

@end

File: Sample Apps/Swift/SampleApp-Swift/MonitoringViewController.swift
//
//  MonitoringViewController.swift
//  SampleApp-Swift
//
//  Created by Nir Lachman on 12/02/2018.
//  Copyright © 2018 LivePerson. All rights reserved.
//

import UIKit
import LPMessagingSDK
import LPInfra
import LPMonitoring

class MonitoringViewController: UIViewController {
    @IBOutlet var accountTextField: UITextField!
    let consumerID = "CONSUMER_ID" // REPLACE THIS!
    let appInstallID = "APP_INSTALL_ID" // REPLACE THIS!
    var pageId: String?
    var campaignInfo: LPCampaignInfo?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: IBActions
    
    /// Init Messaging SDK with brandID (account number) and LPMonitoringInitParams (For monitoring)
    @IBAction func initSDKsClicked(_ sender: Any) {
        defer {
            self.view.endEditing(true)
        }
        
        guard let accountNumber = self.accountTextField.text, !accountNumber.isEmpty else {
            return
        }
        
        let monitoringInitParams = LPMonitoringInitParams(appInstallID: appInstallID)

        do {
            try LPMessagingSDK.instance.initialize(accountNumber, monitoringInitParams: monitoringInitParams)
        } catch let error as NSError {
            print("initialize error: \(error)")
            return
        }
    }
    
    /// Get Engagement clicked selector
    /// Send new Get Engagement using LPMonitoringAPI
    /// CampaignInfo will be saved in the response in order to be able to use routed conversation in Messaging, using the MessagingSDK
    @IBAction func getEngagementClicked(_ sender: Any) {
        defer {
            self.view.endEditing(true)
        }

        let entryPoints = ["tel://972737004000",
                           "http://www.liveperson.com",
                           "sec://Sport",
                           "lang://Eng"]
        let engagementAttributes = [
            ["type": "purchase", "total": 20.0],
            ["type": "lead",
             "lead": ["topic": "luxury car test drive 2015",
                      "value": 22.22,
                      "leadId": "xyz123"]]
        ]

        let monitoringParams = LPMonitoringParams(entryPoints: entryPoints, engagementAttributes: engagementAttributes)
        let identity = LPMonitoringIdentity(consumerID: consumerID, issuer: "")
        LPMonitoringAPI.instance.getEngagement(identities: [identity], monitoringParams: monitoringParams, completion: { [weak self] (getEngagementResponse) in
            print("received get engagement response with pageID: \(String(describing: getEngagementResponse.pageId)), campaignID: \(String(describing: getEngagementResponse.engagementDetails?.first?.campaignId)), engagementID: \(String(describing: getEngagementResponse.engagementDetails?.first?.engagementId))")
            // Save PageId for future reference
            self?.pageId = getEngagementResponse.pageId
            if let campaignID = getEngagementResponse.engagementDetails?.first?.campaignId, let engagementID = getEngagementResponse.engagementDetails?.first?.engagementId, let contextID = getEngagementResponse.engagementDetails?.first?.contextId, let sessionID = getEngagementResponse.sessionId, let visitorID = getEngagementResponse.visitorId {
                self?.campaignInfo = LPCampaignInfo(campaignId: campaignID, engagementId: engagementID, contextId: contextID, sessionId: sessionID, visitorId: visitorID)
            }
        }) { [weak self] (error) in
            self?.campaignInfo = nil
            print("get engagement error: \(error.userInfo.description)")
        }
    }
    
    /// Send SDE clicked selector
    /// Send new SDE using LPMonitoringAPI
    /// PageID in the response will be saved for future request for SDE
    @IBAction func sendSDEClicked(_ sender: Any) {
        defer {
            self.view.endEditing(true)
        }

        let entryPoints = ["http://www.liveperson-test.com",
                           "sec://Food",
                           "lang://De"]
        let engagementAttributes = [
            ["type": "purchase",
             "total": 11.7,
             "orderId": "DRV1534XC"],
            ["type": "lead",
             "lead": ["topic": "luxury car test drive 2015",
                      "value": 22.22,
                      "leadId": "xyz123"]]
        ]

        let monitoringParams = LPMonitoringParams(entryPoints: entryPoints, engagementAttributes: engagementAttributes, pageId: self.pageId)
        let identity = LPMonitoringIdentity(consumerID: consumerID, issuer: "BrandIssuer")
        LPMonitoringAPI.instance.sendSDE(identities: [identity], monitoringParams: monitoringParams, completion: { [weak self] (sendSdeResponse) in
            print("received send sde response with pageID: \(String(describing: sendSdeResponse.pageId))")
            // Save PageId for future reference
            self?.pageId = sendSdeResponse.pageId
        }) { [weak self] (error) in
            self?.pageId = nil
            print("send sde error: \(error.userInfo.description)")
        }
    }
    
    /// Show Conversation clicked selector
    /// Show conversation in MessagingSDK will use the saved CampaignInfo, if received, from the Get Engagement Request
    @IBAction func showConversationWithCampaignClicked(_ sender: Any) {
        guard let campaignInfo = self.campaignInfo, let accountNumber = self.accountTextField.text, !accountNumber.isEmpty  else {
            print("Can't show conversation without valid campaignInfo")
            return
        }

        let conversationQuery = LPMessagingSDK.instance.getConversationBrandQuery(accountNumber, campaignInfo: campaignInfo)
        let conversationViewParam = LPConversationViewParams(conversationQuery: conversationQuery, isViewOnly: false)
        LPMessagingSDK.instance.showConversation(conversationViewParam)
    }
    
    /// Logout clicked selector
    /// Logout Monitoring and Messaging SDKs - all the data will be cleared
    @IBAction func logoutClicked(_ sender: Any) {
        LPMessagingSDK.instance.logout(completion: {
            
        }) { (error) in
            
        }
    }
}


